File: addBoldTag.js
function addBoldTag(s, d) {
    const n = s.length
    const flag = new Array(n)
    let currentEnd = 0
    let ans = ''
    for(let i = 0; i < n; i++) {
        for (let w of d) {
            if (s.startsWith(w, i)) {
                currentEnd = Math.max(currentEnd, i + w.length)
            }
        }
        if (i < currentEnd) {
            flag[i] = true
        } else {
            flag[i] = false
        }
    }
    for(let i = 0; i < n; i++) {
        if(flag[i] && (i == 0 || (i > 0 && !flag[i-1]))) {
            ans += '<b>'
        }
        ans += s[i]
        if (flag[i] && (i==n-1 || (i < n - 1 && !flag[i+1]))) {
            ans += '</b>'
        }
    }
    return ans
}

console.log(addBoldTag('abcxyz123', ['a', '123']))

File: add_bold_tag.py
def addBoldTag(s, d):
    st = [False] * len(s)
    final = ''
    for word in d:
        start = s.find(word)
        end = len(word)
        while start != -1:
            for i in range(start, start + end):
                st[i] = True
            start = s.find(word, start+1)
    i = 0
    while i < len(s):
        if st[i]:
            final += '<b>'
            while i < len(s) and st[i]:
                final += s[i]
                i += 1
            final += '</b>'
        else:
            final += s[i]
            i += 1
    return final

print(addBoldTag('abcxyz123', ['cxy', 'xyz123']))

def addBoldTag(s, d):
    st = [False] * len(s)
    final = ''
    for word in d:
        start = s.find(word)
        end = len(word)
        while start != -1:
            for i in range(start, start + end):
                st[i] = True
            start = s.find(word, start + 1)
    i = 0
    while i < len(s):
        if st[i]:
            final += '<b>'
            while i < len(s) and st[i]:
                final += s[i]
                i += 1
            final += '</b>'
        else:
            final += s[i]
            i += 1
    return final

def addBoldTag(s, words):
    """
    :type s: str
    :type words: List[str]
    :rtype: str
    """
    n = len(s)
    bold = [False] * n
    for word in words:
        start = s.find(word)
        while start != -1:
            for i in range(start, start + len(word)):
                bold[i] = True
            start = s.find(word, start + 1)
    open = '<b>'
    close = '</b>'
    ans = []
    for i in range(n):
        if bold[i] and (i == 0 or not bold[i-1]):
            ans.append(open)
        ans.append(s[i])
        if bold[i] and (i == n - 1 or not bold[i+1]):
            ans.append(close)
    return ''.join(ans)